### PR TITLE
Bug fixes, Refactoring, Tests and new features in exits flow

### PR DESCRIPTION
**Pull request by atvanguard**
- Delete old fraud proofs
- Fix `depositId` 🐛 in `rootChain`
- Fix 🐛🐞 in predicates `verifyDeprecation` logic
- Add a bunch of tests ✅ 
- Major refactoring :recycle:

`WithdrawManager` :construction:
- ~Support multiple exits at the same priority~
- Fix priority queue insertion bug. `queue.insert(exitableAt, priority)` 🚑 
- ~Distinction b/w `priority` and `exitId`~
- New `exitId` format to counter the exits grief-ing vector :art:
```
uint256 ageOfUtxo = (getExitableAt(createdAt) << 127) | (blockNumber << 32) | branchMask.toRlpItem().toUint(); 
uint256 exitId = ageOfUtxo << 1; // The last bit is reserved for housekeeping in erc20Predicate
```
- Add `address adjudicatorPredicate` param in `challengeExit` :building_construction:
This is required because predicate that evaluates the challenge maybe different from the predicate from which the exit was started. For instance, exit may have been started by calling `ERC20Predicate.startExit()` but the challenge could entail invoking `MarketplacePredicate.verifyDeprecation()`.

`Predicates`
- `TransferWithSigPredicate` ✨

- `MarketplacePredicate.verifyDeprecation()` ⚡️ - With this change, marketplace txs become a valid spend i.e. it is possible to challenge an exit with a marketplace tx. Interestingly, this makes a call to `erc20/721Predicate.interpretStateUpdate()` to process the `LogTransfer` receipt.

- Add checks for `transferWithSig` expiration in `MarketplacePredicate.verifyDeprecation()` and `MarketplacePredicate.startExit()`

- Move `address predicate` field from `PlasmaExit` to inside `Input` in `ExitsDataStructure`.

### CODE REVIEW

File: contracts/root/RootChain.sol [hunk 1]
```diff
@@ -57,7 +57,7 @@ contract RootChain is RootChainStorage {
     onlyDepositManager
     returns(uint256 depositId)
   {
-    depositId = _blockDepositId;
+    depositId = _nextHeaderBlock.sub(MAX_DEPOSITS).add(_blockDepositId);
```
Comment (jdkanani) [inline contracts/root/RootChain.sol]: Please explain.

File: contracts/root/withdrawManager/WithdrawManager.sol [hunk 2]
```diff
@@ -82,8 +82,8 @@ contract WithdrawManager is WithdrawManagerStorage, IWithdrawManager {
 
     age = (
       headerNumber * HEADER_BLOCK_NUMBER_WEIGHT +
-      blockNumber * WITHDRAW_BLOCK_NUMBER_WEIGHT +
-      branchMask.toRlpItem().toBytes().toRlpItem().toUint() * BRANCH_MASK_WEIGHT
+      blockNumber * CHILD_BLOCK_NUMBER_WEIGHT +
+      branchMask.toRlpItem().toUint() * BRANCH_MASK_WEIGHT
```
Comment (jdkanani) [inline contracts/root/withdrawManager/WithdrawManager.sol]: This worked?

File: contracts/root/withdrawManager/WithdrawManager.sol [hunk 3]
```diff
@@ -294,22 +291,50 @@ contract WithdrawManager is WithdrawManagerStorage, IWithdrawManager {
...
+  }
+
+  function getExitableWindow()
+    internal
+    returns(uint256)
+  {
+    // Reset exitWindow if it is nearing its end
+    if (now + 1 days >= exitWindow) {
+      exitWindow = now + 1 weeks;
+      exitWindow = now - 1 hours; // JUST FOR TESTING. REMOVE BEFORE MERGING
```
Comment (jdkanani) [inline contracts/root/withdrawManager/WithdrawManager.sol]: Remove if it's not for prod.

File: contracts/root/withdrawManager/WithdrawManager.sol [hunk 4]
```diff
@@ -294,22 +291,50 @@ contract WithdrawManager is WithdrawManagerStorage, IWithdrawManager {
...
+      require(amountOrToken > 0, "CANNOT_EXIT_ZERO_AMOUNTS");
+      key = keccak256(abi.encodePacked(token, exitor));
+    }
+  }
+
+  function getExitableWindow()
+    internal
+    returns(uint256)
+  {
+    // Reset exitWindow if it is nearing its end
```
Comment (jdkanani) [inline contracts/root/withdrawManager/WithdrawManager.sol]: Please explain.

File: contracts/root/RootChain.sol (hunk 1 above)
Comment (atvanguard) [reply contracts/root/RootChain.sol]: If 1 header block was submitted, `_nextHeaderBlock` is 20000.
So `depositId = 20000 - 10000 + 1` where `_blockDepositId` is initialized to 1 at every checkpoint.

If no header blocks have been submitted, `_nextHeaderBlock` is 10000.
So `depositId = 10000 - 10000 + 1`

Basically `_blockDepositId` denotes  deposit count relative to checkpoint and  `depositId = _nextHeaderBlock.sub(MAX_DEPOSITS).add(_blockDepositId)` evaluates absolute depositId.

File: contracts/root/withdrawManager/WithdrawManager.sol (hunk 2 above)
Comment (atvanguard) [reply contracts/root/withdrawManager/WithdrawManager.sol]: It did. The other was buggy infact.

File: contracts/root/withdrawManager/WithdrawManager.sol (hunk 4 above)
Comment (atvanguard) [reply contracts/root/withdrawManager/WithdrawManager.sol]: Discussed offline, not the best solution.

File: contracts/root/withdrawManager/WithdrawManager.sol (hunk 2 above)
Comment (atvanguard) [reply contracts/root/withdrawManager/WithdrawManager.sol]: @0xAshish You were checking this. Please confirm that it works as intended?

**Pull request merged**